telos-backend/src/routes/api/journal.js

```diff
@@ -1,14 +1,119 @@
+// Return a string representation of current day. 
+// See daysOfWeekenum in src/models/habit.js
+function getStringDayOfWeek(dayNum) {
+  let dayString;
+  switch (dayNum) {
+    case 0:
+      dayString = 'sun';
+      break;
+    case 1:
+      dayString = 'mon';
+      break;
+    case 2:
+      dayString = 'tue';
+      break;
+    case 3:
+      dayString = 'wed';
+      break;
+    case 4:
+      dayString = 'thu';
+      break;
+    case 5:
+      dayString = 'fri';
+      break;
+    case 6:
+      dayString = 'sat';
+      break;
+    default:
+      dayString = 'err';
+  }
+  return dayString;
+}
+
+async function getTodosForDay(date){
+  // Todos should return either what is made today, or all overdue tasks.
+  const todos = await TodoTask.find({
+    $or:[
+      { createdDate: { $gte: date + 'T00:00:00', $lte: date + 'T23:59:59' } }, 
+      { dueDate: {$gte: date + 'T00:00:00'} }
+    ] 
+  });
+  return todos;
+}
+
+// Param: date, of type date, parsed from url path param.
+async function getHabitsForDay(date){
+  // Create habit objects to return. These are not stored directly (since technically could be unlimited)
+  // Must find all habits that are between the requested day, but before end date, then check if they fall
+  // on the current day.
+  const habits = await Habit.find({ startDate: {$gte: date + 'T00:00:00'}, endDate: {$lte: date + 'T23:59:59'}});
+  // Habit objects returned in array as following:
+  // {
+  //    name: String
+  //    done: Boolean
+  // }
+  // Note: The Date.getDay() method returns an integer for day of week: 0 for sunday, 1 for monday etc.
+  const day = new Date(date).getDay();
+  const dayOfWeek = getStringDayOfWeek(day)
+  const habitDataToSend = []
+  for (const h in habits){
+    // Check if habit falls on day of week.
+    if (h.daysOfWeek.includes(dayOfWeek)){ 
+      habitDataToSend.push({
+        name: h.name,
+        completed: h.completedDates.includes(new Date(date))
+      })
+    }
+  }
+  return habitDataToSend;
+}
+
+async function getCalendarDataForDay(date) {
+  const calendarevents = await CalendarEvent.find({ startTime: {
+    $gte: date + 'T00:00:00', 
+    $lte: date + 'T23:59:59'
+  }});
+  return calendarevents;
+}
+// ------------------------------------
+// Endpoints defined from this point on 
+
 // get widgets which appear in the journal for a single date.
 // params:
 // 'date': journal date to get widgets for
-router.get('/:date', (req, res) => {
+router.get('/:date', async (req, res) => {
+  const { date } = req.params;
+  //2021-01-01
+  // Need to get all the data from widgets on that date.
+  // Then go through each of components, if exist, and get the data.
+  const widgets = await Widget.find({ date: date });
+
+  // Must find all text widgets based on the widget Ids retrieved.
+  const widgetIds = []
+  for (const w in widgets) {
+    widgetIds.push(w.widgetId);
+  }
+  const texts = await Text.find({ _id: { $in: widgetIds }});
```

Comment:
Are we always going to return all the data for that day regardless of whether there is that widget type on the page? Or we could add a condition outside each function call checking if that widget type is in the list of widgets. I guess the frontend having more data than necessary isn't such a bad thing though, could be better for them???

Reply:
I think for consistency it might be better for now, especially when the amount of data should be reasonably low for a given day. But definitely something we can change in the future. Let me know if you think we should make that change now